hat/assets/js/apps/Iaso/components/forms/AsyncSelect.tsx

```diff
@@ -0,0 +1,167 @@
+import { isArray } from 'lodash';
+
+type Props = {
+    value: any;
+    label: IntlMessage;
+    loading?: boolean;
+    loadingText?: IntlMessage;
+    keyValue: string;
+    // eslint-disable-next-line no-unused-vars
+    onChange: (keyValue, newValue: any | null) => void;
+    errors?: string[];
+    required?: boolean;
+    clearable?: boolean;
+    debounceTime?: number; // debounce time in ms
+    disabled?: boolean;
+    multi?: boolean;
+    helperText?: string;
+    minCharBeforeQuery?: number;
+    // eslint-disable-next-line no-unused-vars
+    fetchOptions: (input: string) => Promise<any[]>;
+};
+
+export const AsyncSelect: FunctionComponent<Props> = ({
+    value,
+    onChange,
+    keyValue,
+    label,
+    loading = false,
+    loadingText = undefined,
+    required = false,
+    clearable = false,
+    debounceTime = 0,
+    disabled = false,
+    multi = false,
+    helperText = undefined,
+    minCharBeforeQuery = 1,
+    fetchOptions,
+}) => {
+    const { formatMessage } = useSafeIntl();
+    const [inputValue, setInputValue] = useState('');
+    const [isLoading, setLoading] = useState(loading);
+    const values = useMemo(() => {
+        if (isArray(value)) {
+            return value;
+        }
+        return value.length > 0
+            ? value.split(',').map(v => {
+                  return {
+                      value: v,
+                      label: v,
+                  };
+              })
+            : [];
+    }, [value]);
+    const [options, setOptions] = useState<readonly any[]>([...values]);
+
+    const fetch = useMemo(
+        () =>
+            debounce(
+                (
+                    request: { input: string },
+                    // eslint-disable-next-line no-unused-vars
+                    callback: (results?: readonly any[]) => void,
+                ) => {
+                    setLoading(true);
+                    fetchOptions(request.input)
+                        .then(newOptions => {
+                            callback(newOptions);
+                            setLoading(false);
+                        })
+                        .catch(() => {
+                            setLoading(false);
+                        });
+                },
+                debounceTime,
+            ),
+        [debounceTime, fetchOptions],
+    );
+
+    useEffect(() => {
+        let active = true;
+        if (inputValue.length < minCharBeforeQuery) {
+            setOptions([...values]);
+            return undefined;
+        }
+
+        fetch({ input: inputValue }, (results?: readonly any[]) => {
+            if (active) {
+                let newOptions: any[] = [...values];
+                if (results) {
+                    newOptions = [...newOptions, ...results];
+                }
+                setOptions(newOptions);
+            }
+        });
+        return () => {
+            active = false;
+        };
+    }, [values, inputValue, fetch]);
```

Comment:
i don't really get this `useEffect`, if the purpose is to get options, maybe calculate the options in a `useMemo` instead ?

Reply:
If my understanding is correct, `useMemo` needs something to return. In this case, we want to execute something as long as the component is active (therefore, the `active` boolean) when the value of the input changes.
Also, I took this from here: https://mui.com/material-ui/react-autocomplete/#search-as-you-type